### PR TITLE
feat: BUILDER_PLUGIN provide a cache file for skipping building pkgs

### PR DESCRIPTION
**Pull request by kalkin**
The `changed-files` script makes sure that the `CACHE_FILE` target is
rebuild each time the source code changes.

### CODE REVIEW

**Comment by kalkin** [comment]
See also the corresponding pr for https://github.com/QubesOS/qubes-builder-rpm/pull/94

File: Makefile.generic [hunk 1]
```diff
@@ -208,8 +207,21 @@ packages:
...
+ifneq (,$(PACKAGE))
 .PHONY: package
-package: dist-build-dep dist-package dist-copy-out $(SOURCE_COPY_OUT)
+ifeq (,$(CACHE_FILE))
+$(info  -> Building $(COMPONENT) $(PACKAGE) for $(DIST)-$(PACKAGE_SET))
+package: | dist-build-dep dist-package dist-copy-out $(SOURCE_COPY_OUT)
+else
+$(CACHE_FILE) : $(addprefix $(ORIG_SRC)/,$(shell scripts/changed-files $(ORIG_SRC)))
+       $(info  -> Rebuilding packages for $(COMPONENT)/pkgs/$(PACKAGE_SET)-$(DIST) from $(PACKAGE))
+       rm $@
```
Comment (marmarek) [inline Makefile.generic:218]: `rm -f` (to not fail on a missing file)

File: Makefile.generic [hunk 2]
```diff
@@ -208,8 +207,21 @@ packages:
 		fi;\
 	done
 
+ifneq (,$(PACKAGE))
 .PHONY: package
-package: dist-build-dep dist-package dist-copy-out $(SOURCE_COPY_OUT)
+ifeq (,$(CACHE_FILE))
```
Comment (marmarek) [inline Makefile.generic:212]: I'd like to avoid too many conditionals if possible. And I think it is possible here. Those conditionals (on CACHE_FILE and on PACKAGE) seems to be needed only because CACHE_FILE is set by a builder plugin conditionally (if the builder plugin supports it, and if PACKAGE is set). But I think builder plugin doesn't need to be involved here. What you need is a build timestamp to compare with source files, it doesn't need to have any specific content. It would be simpler to create a build stamp file here (independent of existing .list) based on variables you already have. Bonus - it will work without any builder plugin change, which means you have it for all the package types for free.

Besides the above, I don't like `CACHE_FILE` name - it suggests some content is actually cached inside, plus it may be confused with `CACHEDIR`. `BUILD_STAMP_FILE`? 

So, it could look like this (untested):
```
BUILD_STAMP_FILE = $(ORIG_SRC)/$(OUTPUT_DIR)/$(notdir $(PACKAGE).build_stamp
$(BUILD_STAMP_FILE) : $(addprefix $(ORIG_SRC)/,$(shell scripts/changed-files $(ORIG_SRC)))
    $(info  -> Rebuilding packages for $(COMPONENT)/pkgs/$(PACKAGE_SET)-$(DIST) from $(PACKAGE))
    $(MAKE) -f $(THIS_MAKEFILE) PACKAGE=$(PACKAGE) dist-build-dep dist-package dist-copy-out $(SOURCE_COPY_OUT)
    touch "$@"

package: ...
```

File: scripts/changed-files [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+set -e
+[ $# -ne 1 ] && echo "Usage: $0 DIR" >&2 && exit 1
+cd "$1"
+STAGED_UNSTAGED=$(git diff --name-only HEAD)
+[ -n "$STAGED_UNSTAGED" ] && echo "$STAGED_UNSTAGED" && exit
```
Comment (marmarek) [inline scripts/changed-files:6]: Why this part? Doesn't `git ls-files` handle this case too? In fact, if this isn't needed, the whole script can be replaced with `git -C ... ls-files`,

**Review by marmarek — CHANGES_REQUESTED** [review]
I like the simplicity of this solution in general!
I haven't tested it, but looks promising. In fact, I think it can be simplified even further - see specific comments below.

File: Makefile.generic (hunk 1 above)
Comment (kalkin) [reply Makefile.generic]: Fixed

File: Makefile.generic (hunk 2 above)
Comment (kalkin) [reply Makefile.generic]: Done, but I moved the whole rebuilding logic to own Goal. This has the effect that the order of `|dist-build-dep dist-package…` Prereqs is preserved. This will avoid issues once I fill `qubes-builder` to be able to run with `make -j`

File: scripts/changed-files (hunk 3 above)
Comment (kalkin) [reply scripts/changed-files]: Premature optimization…

**Comment by kalkin** [comment]
This solution has one issue: It still runs `dist-prep-chroot` which takes some time. I think we need another `BUILD_STAMP` in the main `Makefile` for each `COMPONENT`

**Comment by marmarek** [comment]
Maybe move `dist-prepare-chroot` to per-package call? Most components have a single package, so it won't make big difference.

**Comment by kalkin** [comment]
> Maybe move dist-prepare-chroot to per-package call

Why is `dist-prepare-chroot` even exposed and not part of inernal `BUILDR_PLUGIN` machinery?

**Comment by marmarek** [comment]
> Why is `dist-prepare-chroot` even exposed and not part of inernal `BUILDR_PLUGIN` machinery?

Mostly for CI which caches pristine prepared chroots: https://github.com/QubesOS/qubes-continuous-integration/blob/master/r4.1/gitlab-base.yml#L65

**Comment by kalkin** [comment]
Do not merge. 

Note to myself: This PR is missing a fix for cleaning rpms. Extract from my primore clean up and add to `qb` upstream
```
@find qubes-src -path \*/pkgs/\*.build_stamp -delete -print | sed -e 's/^/ removed /'
```

**Comment by kalkin** [comment]
Lol the further it goes down the rabbit hole the more errors I find. 

I think we need a  `changed-files` script in some way. Creating the stampfile for `installer-qubes-os` fails. `git ls-files` returns the file `lorax-templates-qubes/templates/config_files/x86/System Volume Information/IndexerVolumeGuid`, which breaks `make(1)` because of whitespace.

**Comment by kalkin** [comment]
Next issue. As far as I see `make iso` calls `update-repo` for `template.spec`. I changed `update-repo` to work completly from  the `*.list` file instead of querying the spec files, but `make template` does not create a `BUILD_STAMP_FILE` or `*.list` file. 

I'm not sure what is the right way to solve it:

* Add exceptions to Makefile.generic (more `ifeq/else/endif` :vomiting_face:)
* Change `make template` to generate needed files and to put the template rpm packages in `/pkg/...` dir instead of `/rpm`.
* Change `update-repo` back how it was

Generally currently building the template RPM package does not use the caching code. I think it should. I'm also not sure why `make template` and `make iso` first clean up all packages and then copy them back. IMHO this should be handled by make and proper `prereqs` to save work.

**Comment by marmarek** [comment]
> Change `make template` to generate needed files and to put the template rpm packages in `/pkg/...` dir instead of `/rpm`.

I think this is the best option. It sounds also quite easy to implement.

> Generally currently building the template RPM package does not use the caching code. I think it should.

That isn't this obvious. Often the template rebuild happens not because of some qubes package change, but to pick up newer base distribution packages.

> I'm also not sure why `make template` and `make iso` first clean up all packages and then copy them back. IMHO this should be handled by make and proper `prereqs` to save work.

This is to ensure no leftovers from previous build (especially failed one). Including cases when you add or remove something from COMPONENTS, or switch branch which may pull in an older version (so - leftover newer version would clobber it). Such failure mode is really frustrating to debug, because in most cases the actual iso/template build will be successful but it would include wrong packages, which may be hard to spot. I think a bit less optimal build process is a good compromise to hours wasted due to such issues...

**Comment by marmarek** [comment]
> Creating the stampfile for `installer-qubes-os` fails. `git ls-files` returns the file `lorax-templates-qubes/templates/config_files/x86/System Volume Information/IndexerVolumeGuid`, which breaks `make(1)` because of whitespace.

can `make(1)` handle files with whitespaces at all? if so, maybe there is a way to correctly feed `git ls-files` output to it. When doing a cleanup like this, I'd like to not go back to external scripts for checking if rebuild is needed and then `if...` this in makefile - generally conditionals like this in the makefile makes it hard to follow and almost always incompatible with `make -j`.

If external script is not avoidable, then maybe touch some file with a date of the most recent file, and then have make compare it against the build stamp to decide about the rebuild?

**Comment by kalkin** [comment]
> > Creating the stampfile for `installer-qubes-os` fails. `git ls-files` returns the file `lorax-templates-qubes/templates/config_files/x86/System Volume Information/IndexerVolumeGuid`, which breaks `make(1)` because of whitespace.
> 
> can `make(1)` handle files with whitespaces at all?

Yes I found a sollution but it's hacky `git ls-files|sed -e 's# #\\ #g'`

**Comment by kalkin** [comment]
> I think this is the best option. It sounds also quite easy to implement.

kthx

**Comment by kalkin** [comment]
There is something weird going on when building `linux-kernel`
```
make[1]: Entering directory '/home/user/qubes-builder-back'
-> Building linux-kernel (kernel.spec) for fc32 dom0 (logfile: build-logs/linux-kernel-dom0-fc32.log)
make[2]: Entering directory '/home/user/qubes-builder-back'
-> Building packages for linux-kernel/pkgs/dom0-fc32 from kernel.spec
rm -f qubes-src/linux-kernel/pkgs/dom0-fc32/kernel.spec.build_stamp
make -f Makefile.generic PACKAGE=kernel.spec rebuild-package
make[3]: Entering directory '/home/user/qubes-builder-back'
fatal: cannot change to 'qubes-src/linux-kernel/kernel.spec': No such file or directory
fatal: cannot change to 'qubes-src/linux-kernel/kernel.spec': No such file or directory
fatal: cannot change to 'qubes-src/linux-kernel/kernel.spec': No such file or directory
/home/user/qubes-builder-back/scripts/create-archive /home/user/qubes-builder-back/chroot-dom0-fc32//home/user/qubes-src/linux-kernel/kernel.spec kernel.spec-.tar.gz kernel.spec/
fatal: cannot change to 'qubes-src/linux-kernel/kernel.spec': No such file or directory
/home/user/qubes-builder-back/scripts/create-archive: line 22: pushd: /home/user/qubes-builder-back/chroot-dom0-fc32/home/user/qubes-src/linux-kernel/kernel.spec: No such file or directory
make[3]: *** [qubes-src/linux-kernel/Makefile.builder:11: dummy-psu] Error 1
make[3]: Leaving directory '/home/user/qubes-builder-back'
make[2]: *** [Makefile.generic:221: qubes-src/linux-kernel/pkgs/dom0-fc32/kernel.spec.build_stamp] Error 2
make[2]: Leaving directory '/home/user/qubes-builder-back'
--> build failed!
make[1]: *** [Makefile.generic:193: packages] Error 1
make[1]: Leaving directory '/home/user/qubes-builder-back'
make: *** [Makefile:262: linux-kernel-dom0] Error 1
```

I'm kind of confused by the error

**Comment by marmarek** [comment]
Does the `qubes-src/linux-kernel` directory exist at all?

**Comment by kalkin** [comment]
> Does the `qubes-src/linux-kernel` directory exist at all?

Yes

**Comment by kalkin** [comment]
On master the linux-kernel starts building like it should

**Comment by kalkin** [comment]
I think i found the culprit. On `master` the `linux-kernel` `Makefile.builder` is called with the following arguments for the variable `PACKAGE`:
* dummy-psu
* dummy-backlight
* linux-utils

On my branch it's called with `kernel.spec` . Did I miss some `ifdefs` for `linux-kernel` handling?

**Comment by marmarek** [comment]
This looks very suspicious, likely variable name conflict. Take a look at `linux-kernel/Makefile.builder`.

**Comment by marmarek** [comment]
I've changed variable name in the linux-kernel repo. Does it work now?

**Comment by kalkin** [comment]
Same issue in`vmm-xen-stubdom-legacy`, once I fixed it, it picked up building where I left it a few weeks ago, like it supposed to. Will see if any new errors arise

**Comment by kalkin** [comment]
`mgmt-salt` fails with:
```
-> Building mgmt-salt (rpm_spec/qubes-mgmt-salt.spec) for fc32 dom0 (logfile: build-logs/mgmt-salt-dom0-fc32.log)
--> build failed!
make[2]: *** No rule to make target 'qubes-src/mgmt-salt/tests/00-packages-go-in-here-for-fc20', needed by 'qubes-src/mgmt-salt/pkgs/dom0-fc32/qubes-mgmt-salt.spec.build_stamp'.  Stop.
```

After executing `git rm qubes-src/mgmt-salt/tests/00-* && git commit` the build continued. The commit was needed, because make expects the files to be there because of previous `git ls-files`

**Comment by kalkin** [comment]
Building templates fails for some completly different reason:
```
└─rpmfactory:✘> make template VERBOSE=1
Building template: fedora-32
-> Preparing instalation of fc32 template...
-> Image file already exists, assuming *update*...
INFO: /home/user/qubes-builder-back/qubes-src/builder-rpm/template_scripts/distribution.sh imported by: /home/user/qubes-builder-back/qubes-src/builder-rpm/template_scripts/01_install_core.sh
Unable to detect release version (use '--releasever' to specify release version)
Error: Failed to download metadata for repo 'fedora-modular': Cannot prepare internal mirrorlist: Status code: 404 for https://mirrors.fedoraproject.org/metalink?repo=fedora-modular-$releasever&arch=x86_64 (IP: 152.19.134.198)
Attempting to kill any processes still running in '/home/user/qubes-builder-back/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/qubes-builder-back/qubes-src/linux-template-builder/mnt/tmp/template-builder-repo
INFO: umount /home/user/qubes-builder-back/qubes-src/linux-template-builder/mnt/etc/resolv.conf
INFO: umount /home/user/qubes-builder-back/qubes-src/linux-template-builder/mnt
make[1]: *** [Makefile:65: rootimg-build] Error 1
make: *** [Makefile:341: template-local-fc32] Error 1
```

I will try to patch it locally, but I'm surprised.

**Comment by kalkin** [comment]
Building templates fails reproducible with name resolution error. May be the chroot/mock network is somehow broken?
```
+ /usr/sbin/chroot /home/user/qubes-builder-back/qubes-src/linux-template-builder/mnt dnf --downloadonly --releasever=33 install -y /tmp/3rd_party_software/adobe-release-x86_64.noarch.rpm
Qubes Template Builder Repository                                                                           3.7 MB/s | 3.8 kB     00:00
Fedora 33 openh264 (From Cisco) - x86_64                                                                    0.0  B/s |   0  B     00:40
Errors during downloading metadata for repository 'fedora-cisco-openh264':
  - Curl error (6): Couldn't resolve host name for https://codecs.fedoraproject.org/openh264/33/x86_64/os/repodata/fa9f31c08821f824e9371eeeb918aa00385d9d865ccfd6d54fb996dcc50a6830-primary.xml.gz [Could not resolve host: codecs.fedoraproject.org]
  - Curl error (6): Couldn't resolve host name for https://codecs.fedoraproject.org/openh264/33/x86_64/os/repodata/47753f30191e677ffc6d20cc5baf9eeec85e6c9b09b4ee5d7fe4c09ebea4f372-filelists.xml.gz [Could not resolve host: codecs.fedoraproject.org]
  - Curl error (6): Couldn't resolve host name for https://codecs.fedoraproject.org/openh264/33/x86_64/os/repodata/3d50825fb25440abc9579f078205d6ba3590b1d146187af8179eb43214c8b40e-comps-Temporary.x86_64.xml.gz [Could not resolve host: codecs.fedoraproject.org]
Error: Failed to download metadata for repo 'fedora-cisco-openh264': Yum repo downloading error: Downloading error(s): repodata/fa9f31c08821f824e9371eeeb918aa00385d9d865ccfd6d54fb996dcc50a6830-primary.xml.gz - Cannot download, all mirrors were already tried without success; repodata/47753f30191e677ffc6d20cc5baf9eeec85e6c9b09b4ee5d7fe4c09ebea4f372-filelists.xml.gz - Cannot download, all mirrors were already tried without success; repodata/3d50825fb25440abc9579f078205d6ba3590b1d146187af8179eb43214c8b40e-comps-Temporary.x86_64.xml.gz - Cannot download, all
mirrors were already tried without success
google-chrome - x86_64                                                                                       10 kB/s | 3.6 kB     00:00
Qubes OS Repository for VM (updates)                                                                        1.7 kB/s | 833  B     00:00
Qubes OS Repository for VM (updates)                                                                        2.2 MB/s | 2.3 kB     00:00
Importing GPG key 0x9E2795E9:
 Userid     : "Qubes OS Release 4 Signing Key"
 Fingerprint: 5817 A43B 283D E5A9 181A 522E 1848 792F 9E27 95E9
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-qubes-4-primary
Qubes OS Repository for VM (updates)                                                                        0.0  B/s |   0  B     00:40
Errors during downloading metadata for repository 'qubes-vm-r4.1-current':
  - Curl error (6): Couldn't resolve host name for https://yum.qubes-os.org/r4.1/current/vm/fc33/repodata/0a4d9ec5911f793f87955cf8bbc866b0c930c8f9aa2384ec29cc1f360a31da93-comps.xml.gz [Could not resolve host: yum.qubes-os.org]
  - Curl error (6): Couldn't resolve host name for https://yum.qubes-os.org/r4.1/current/vm/fc33/repodata/f2711ab493bede1af94627284a43363f9b67cf02ddc172101273a9e32d642e6a-primary.xml.gz [Could not resolve host: yum.qubes-os.org]
  - Curl error (6): Couldn't resolve host name for https://yum.qubes-os.org/r4.1/current/vm/fc33/repodata/b22cc4970546215fb991bb28d5977e30743c0929907922c014c5805f05e7b4f0-filelists.xml.gz [Could not resolve host: yum.qubes-os.org]
Error: Failed to download metadata for repo 'qubes-vm-r4.1-current': Yum repo downloading error: Downloading error(s): repodata/f2711ab493bede1af94627284a43363f9b67cf02ddc172101273a9e32d642e6a-primary.xml.gz - Cannot download, all mirrors were already tried without success; repodata/b22cc4970546215fb991bb28d5977e30743c0929907922c014c5805f05e7b4f0-filelists.xml.gz - Cannot download, all mirrors were already tried without success; repodata/0a4d9ec5911f793f87955cf8bbc866b0c930c8f9aa2384ec29cc1f360a31da93-comps.xml.gz - Cannot download, all mirrors were already tried without success
+ retval=1
+ true
```

**Comment by kalkin** [comment]
I can reproduce the issue time and time again. For some reason it only happens at this step, before it has no issues connecting to the qubes repository

**Comment by kalkin** [comment]
I'm lost here. This is definitely not an issue with my network connection.

**Comment by marmarek** [comment]
Two ideas (possibly both issues):
 - rpmdb format version - previously all packages we installed via `chroot ... dnf ...`, now initial phase is done with `dnf --installroot=...`, which may produce mismatching rpmdb version. Theoretically it [is re-generated at some point](https://github.com/QubesOS/qubes-builder-rpm/commit/6ea6b19a943fc2177e5462d47bb8dfb5ad1c90b7#diff-8b48ce9d9fad5556c63e4e4bf1cbf3c9e34fbf51b6ed831bc8b1fa377deb9b7eR9-R17), but maybe there is some issue with this code
 - missing/invalid /etc/resolv.conf inside the chroot (again, may be related to the above change)

**Comment by kalkin** [comment]
@marmarek Looking at the `grep -v openSUSE` line, is this a workaround for suse only or should it have been`grep -qv openSUSE`?

**Comment by marmarek** [comment]
I think even more - it should be applied always. I think this specific conversion handles migration into sqlite(?) rpmdb format, which landed in openSUSE earlier than in Fedora - that's why it was under this condition before.

**Comment by marmarek** [comment]
That said, I'm not sure if that's really the source of the issue you see.

**Comment by kalkin** [comment]
I think i found the issue, but I'm not sure what would be the correct way to fix it. The `/etc/resolv.conf` file is blank:
```
+ cat /home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt/etc/resolv.conf
# This file intentionally left blank
```

**Comment by marmarek** [comment]
`/etc/resolv.conf` should be bind-mounted from the host: https://github.com/QubesOS/qubes-builder-rpm/blob/master/template_scripts/distribution.sh#L83
It seems for some reason it doesn't work. Maybe your (real) `/etc/resolv.conf` is a symlink instead of a normal file?

**Comment by kalkin** [comment]
@marmarek Ohh, I see. My previous assumption is invalid, because I did a `cat` to early in the script.

**Comment by kalkin** [comment]
The content of `/etc/resolv.conf` seems to be right up to the point where it imports the gpg key. After that the download fails. I suspect there is a call to dnf which doesn't execute a `mount --bind` before.

```
DEBUG: Smart files located for: 'packages_qubes_+.list' (suffix: ):
DEBUG: /home/user/Projects/qubes-builder/qubes-src/builder-rpm/template_scripts/packages_qubes.list
DEBUG: Installing extra packages from: /home/user/Projects/qubes-builder/qubes-src/builder-rpm/template_scripts/packages_qubes.list
INFO: Packages: qubes-vm-dependencies qubes-vm-recommended
<RESOLV_CONF>
nameserver 10.139.1.1
nameserver 10.139.1.2
</RESOLV_CONF>
Importing GPG key 0x9E2795E9:
 Userid     : "Qubes OS Release 4 Signing Key"
 Fingerprint: 5817 A43B 283D E5A9 181A 522E 1848 792F 9E27 95E9
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-qubes-4.1-primary
Error: Failed to download metadata for repo 'qubes-builder-vm-r4.1-current': Yum repo downloading error: Downloading error(s): repodata/f2711ab493bede1af94627284a43363f9b67cf02ddc172101273a9e32d642e6a-primary.xml.gz - Cannot download, all mirrors were
already tried without success; repodata/b22cc4970546215fb991bb28d5977e30743c0929907922c014c5805f05e7b4f0-filelists.xml.gz - Cannot download, all mirrors were already tried without success; repodata/0a4d9ec5911f793f87955cf8bbc866b0c930c8f9aa2384ec29cc1f360a31da93-comps.xml.gz - Cannot download, all mirrors were already tried without success
Attempting to kill any processes still running in '/home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt' before un-mounting
INFO: umount /home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt/tmp/template-builder-repo
INFO: umount /home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt/proc
INFO: umount /home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt/etc/resolv.conf
INFO: umount /home/user/Projects/qubes-builder/qubes-src/linux-template-builder/mnt
make[1]: *** [Makefile:66: rootimg-build] Error 1
make: *** [Makefile:348: template-local-fc33] Error 1
```

**Comment by kalkin** [comment]
The `/etc/resolv.conf` seems to be correct, but it still fails to resolve.

**Comment by marmarek** [comment]
The last one doesn't look like resolv.conf issue, it is failure to import repo metadata key. Check if the key mentioned in the message is really there.

**Comment by kalkin** [comment]
> The last one doesn't look like resolv.conf issue, it is failure to import repo metadata key. 
Is it? For me it looks like the key exists, or else it wouldn't print out the key data before failing. 

> Check if the key mentioned in the message is really there.

Looks like it exists.
```
$ fd RPM-GPG-KEY-qubes-4.1-primary
qubes-builder/qubes-src/builder-rpm/keys/RPM-GPG-KEY-qubes-4.1-primary
qubes-builder/qubes-src/qubes-release/RPM-GPG-KEY-qubes-4.1-primary
```

**Comment by kalkin** [comment]
The culprit may actually be the `chroot_cmd sh -c 'rpm --import…'` command in `04_install_qubes.sh`. If my assumption is correct, this happens because the usual `mount --bind` does not happen, like during `installPackages()` call. I commented out the line, we will see if the error still occurs.

**Comment by kalkin** [comment]
The error and also the import still happen. May be it's the wrong place?

**Comment by kalkin** [comment]
@marmarek Any ideas?

**Comment by kalkin** [comment]
Because I can reproduce the error with the stock `qubes-builder` at exactly the same step (in fc32 & fc33 templates) I assume it's some template builder error.

Am I currently the only one which can't run the target `template` successfully?
